cli/lsp/repl.rs

```diff
@@ -0,0 +1,292 @@
+  cwd_uri: ModuleSpecifier,
+}
+
+impl ReplLanguageServer {
+  pub async fn new_initialized() -> Result<ReplLanguageServer, AnyError> {
+    super::logging::set_lsp_log_level(log::Level::Debug);
+    let language_server =
+      super::language_server::LanguageServer::new(Client::new_for_repl());
+
+    let cwd_uri = get_cwd_uri()?;
+
+    #[allow(deprecated)]
+    language_server
+      .initialize(InitializeParams {
+        process_id: None,
+        root_path: None,
+        root_uri: Some(cwd_uri.clone()),
+        initialization_options: Some(
+          serde_json::to_value(get_repl_workspace_settings()).unwrap(),
+        ),
+        capabilities: ClientCapabilities {
+          workspace: None,
+          text_document: None,
+          window: None,
+          general: None,
+          experimental: None,
+        },
+        trace: None,
+        workspace_folders: None,
+        client_info: Some(ClientInfo {
+          name: "Deno REPL".to_string(),
+          version: None,
+        }),
+        locale: None,
+      })
+      .await?;
+
+    language_server.initialized(InitializedParams {}).await;
+
+    let server = ReplLanguageServer {
+      language_server,
+      document_version: 0,
+      document_text: String::new(),
+      pending_text: String::new(),
+      cwd_uri,
+    };
+    server.open_current_document().await;
+
+    Ok(server)
+  }
+
+  pub async fn commit_text(&mut self, line_text: &str) {
+    self.did_change(line_text).await;
+    self.document_text.push_str(&self.pending_text);
+    self.pending_text = String::new();
+  }
+
+  pub async fn completions(
+    &mut self,
+    line_text: &str,
+    position: usize,
+  ) -> Vec<ReplCompletionItem> {
+    self.did_change(line_text).await;
+    let before_line_len = BytePos(self.document_text.len() as u32);
+    let position = before_line_len + BytePos(position as u32);
+    let text_info = deno_ast::SourceTextInfo::from_string(format!(
+      "{}{}",
+      self.document_text, self.pending_text
+    ));
+    let line_and_column = text_info.line_and_column_index(position);
+    let response = self
+      .language_server
+      .completion(CompletionParams {
+        text_document_position: TextDocumentPositionParams {
+          text_document: TextDocumentIdentifier {
+            uri: self.get_document_specifier(),
+          },
+          position: Position {
+            line: line_and_column.line_index as u32,
+            character: line_and_column.column_index as u32,
+          },
+        },
+        work_done_progress_params: WorkDoneProgressParams {
+          work_done_token: None,
+        },
+        partial_result_params: PartialResultParams {
+          partial_result_token: None,
+        },
+        context: None,
+      })
+      .await
+      .ok()
+      .unwrap_or_default();
+
+    let items = match response {
+      Some(CompletionResponse::Array(items)) => items,
+      Some(CompletionResponse::List(list)) => list.items,
```

Comment:
discussion: so for v2 of the registry, lists can come back with `isIncomplete` set to `true`.  This is a flag to indicate to the client that not everything is returned and the further queries are needed to refine the results.  In v2 version of the registry, this is how a registry can provide folder by folder "completion" of a URL.  The vscode client uses it to know if there is any value in querying the server again and does so in a "debounced" way.  If it is `isIncomplete` set to `false`, the server is saying you have the definitive result set and should "filter" it on the client.

Reply:
I think for now it shouldn't matter for the repl because it will just ask for completions whenever the user hits tab and doesn't do completions when typing characters or anything. Probably once we start doing completions like shown then we can come back to this:

![image](https://user-images.githubusercontent.com/1609021/146090782-6b1a5a08-8cf9-4d42-bda0-d732fe4abd70.png)